mllib/src/main/scala/org/apache/spark/mllib/clustering/DpMeans.scala

```diff
@@ -0,0 +1,247 @@
+        axpy(1.0, x._1, y._1)
+        (y._1, x._2 + y._2)
+      }
+
+      // Loop until all data points are covered by some cluster center
+      do {
+        localCenters = zippedData.mapPartitions(h => DpMeans.cover(h, globalCenters, lambda))
+          .collect()
+        if (localCenters.isEmpty) {
+          covered = true
+        }
+        // Promote a local cluster center to a global center
+        else {
+          var newGlobalCenters = DpMeans.cover(localCenters.iterator,
+            ArrayBuffer.empty[VectorWithNorm], lambda)
+          globalCenters ++= newGlobalCenters
+        }
+      } while (covered == false)
+
+      // Find the sum and count of points belonging to each cluster
+      val clusterStat = zippedData.mapPartitions { points =>
+        val activeCenters = globalCenters
+        val k = activeCenters.length
+        val dims = activeCenters(0).vector.size
+
+        val sums = Array.fill(k)(Vectors.zeros(dims))
+        val counts = Array.fill(k)(0L)
+        val totalCost = Array.fill(k)(0.0D)
+
+        points.foreach { point =>
+          val (currentCenter, cost) = DpMeans.assignCluster(activeCenters, point)
+          totalCost(currentCenter) += cost
+          val currentSum = sums(currentCenter)
+          axpy(1.0, point.vector, currentSum)
+          counts(currentCenter) +=1
+        }
+
+        val result = for (i <- 0 until k) yield {
+          (i, (sums(i), counts(i)))
+        }
+        result.iterator
+      }.reduceByKey(mergeClusters).collectAsMap()
+
+      // Update the cluster centers
+      var changed = false
+      var j = 0
+      val currentK = clusterStat.size
+      while (j < currentK) {
+        val (sumOfPoints, count) = clusterStat(j)
+        if (count != 0) {
+          scal(1.0 / count, sumOfPoints)
+          val newCenter = new VectorWithNorm(sumOfPoints)
+          // Check for convergence
+          globalCenters.length match {
+            case currentK => if (DpMeans.squaredDistance(newCenter, globalCenters(j)) >
+              convergenceTol * convergenceTol) {
+              changed = true
+              }
+            case _ => changed = true
+          }
+          globalCenters(j) = newCenter
+        }
+        j += 1
+      }
+      if (!changed) {
+        converged = true
+        logInfo("DpMeans clustering finished in " + (iteration + 1) + " iterations")
+      }
+      iteration += 1
+      norms.unpersist()
+    }
+
+    if (iteration == maxIterations) {
+      logInfo(s"DPMeans reached the max number of iterations: $maxIterations.")
+    } else {
+      logInfo(s"DPMeans converged in $iteration iterations")
+    }
+    new DpMeansModel(globalCenters.toArray.map(_.vector))
+  }
+}
+/**
+ * Core methods of  DP means clustering.
+ */
+private object DpMeans {
+  /**
+   * A data point is said to be "covered" by a cluster `c` if the distance from the point
+   * to the cluster center of `c` is less than a given lambda value.
```

Comment:
Missing doc for params and, more importantly, the return type. The API doc reads like that the method returns a boolean instead.